python/lsst/daf/butler/datastores/genericDatastore.py

```diff
@@ -0,0 +1,169 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Generic datastore code useful for most datastores."""
+
+__all__ = ("GenericBaseDatastore", )
+
+import logging
+from typing import MutableMapping
+
+from lsst.daf.butler import Datastore, DatasetTypeNotSupportedError
+
+log = logging.getLogger(__name__)
+
+
+class GenericBaseDatastore(Datastore):
+    """Methods useful for most implementations of a `Datastore`.
+
+    Should always be sub-classed since key abstract methods are missing.
+    """
+
+    records: MutableMapping
+    """Place to store internal records about datasets."""
```

Comment:
If this is only used in the default implementation of `removeStoredItemInfo`, I think it might be cleaner overall to leave that abstract (like `addStoredItemInfo`) and let subclasses define `records` or not; it might avoid some code duplication today, but I don't think `DatabaseDict` will survive adding vectorized inserts of datasets, and that I think that means we shouldn't expect the records for `InMemoryDatastore` to continue to look like the records for other `Datastores`.